SPARK: Remove dependency on hadoop's filesystem class from remove orphan files

liziyan-lzy opened the pull request:
Closes #11541  
RemoveOrphanFiles Using FileIO when it implements SupportPrefixOperations to list files. And it will fall back using Hadoop Class when it does not.

Some concerns:
1. All list file operations are performed on the driver, which may cause performance pressure in scenarios involving massive files.
2. Currently, it has only been implemented in Spark 3.5. Should it be implemented in other Spark versions as well?
3. Tests are running on this FileIO branch. Should tests be added for the original Hadoop Class based branch?

liziyan-lzy commented:
@RussellSpitzer  Hi, please help me review this when you have time. Thx.

eshishki commented:
@liziyan-lzy i think there is a problem with listprefix as is
while removing orphan files for table invoices it also removed files from table invoices_other

liziyan-lzy commented:
> @liziyan-lzy i think there is a problem with listprefix as is while removing orphan files for table invoices it also removed files from table invoices_other

@eshishki Thanks for your review. I'm sorry, but I didn't fully understand what you meant. Could you please explain it in a bit more detail? Thanks.

eshishki commented:
@liziyan-lzy i have 2 tables
invoices and invoices_payments
i run remove_orphane_files('invoices')
this patch uses listprefix api to search for files for invoices tables
basically it returns data for invoices* and deletes files for invoices_payments as they are not part of invoices manifests

i fixed it by passing location option with invoices/, slash here is important
i think it should be done by default

liziyan-lzy commented:
> @liziyan-lzy i have 2 tables invoices and invoices_payments i run remove_orphane_files('invoices') this patch uses listprefix api to search for files for invoices tables basically it returns data for invoices* and deletes files for invoices_payments as they are not part of invoices manifests
> 
> i fixed it by passing location option with invoices/, slash here is important i think it should be done by default

Thanks for pointing this out!
You’re right. I’ve fixed it now.
Let me know if you spot anything else or have further suggestions. Much appreciated!

pvary commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/DeleteOrphanFilesSparkAction.java:
```diff
@@ -302,21 +303,29 @@ private Dataset<FileURI> actualFileIdentDS() {
+      Predicate<FileStatus> predicate = file -> file.getModificationTime() < olderThanTimestamp;
+      // list at most MAX_DRIVER_LISTING_DEPTH levels and only dirs that have
+      // less than MAX_DRIVER_LISTING_DIRECT_SUB_DIRS direct sub dirs on the driver
+      listDirRecursivelyWithHadoop(
```
Will we get here with HadooFileIO?
AFAIK the `HadooFileIO` implements `SupportsPrefixOperations`

liziyan-lzy replied:
Thanks for your review. HadoopFileIO will not run to this branch. This implementation is mainly for backward compatibility. Handle some scenarios where the FileSystem schema is loaded via Hadoop configuration.

pvary replied:
So basically if the FileIO doesn't implement the `SupportsPrefixOperations` (like OSS, ECS), then we will still try to list the files through Hadoop libraries, but not through the HadoopFileIO.

It is good to have this for backward compatibility, but if we don't have tests for this codepath, then we might cause unintended issues later. So please create tests which exercise this codepath

liziyan-lzy replied:
Thanks for pointing this out. You're right. We should test the fallback path. I'll add tests to cover this scenario and ensure backward compactibility works as expected.

RussellSpitzer replied:
You could just add to the code above that if the FS is HadoopFileIO, just use the old path. Other alternatives would be to add a flag, so you only use the new path when the flag is set to true. That's probably a safe migration path as well

pvary replied:
I think, I would prefer a flag. I still feel that we might face performance issues for big tables with many files, and having an option to fall back to the original code-path would be a safe way to move forward.

liziyan-lzy replied:
I have added a flag to switch between old (default) and new paths. Tests covers two paths. Thanks.

pvary commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/DeleteOrphanFilesSparkAction.java:
```diff
@@ -335,7 +344,39 @@ private Dataset<String> listedFileDS() {
+    String listPath = dir;
+    if (!dir.endsWith("/")) {
+      listPath = dir + "/";
+    }
```
nit: newline

liziyan-lzy replied:
Fixed. Thanks!

pvary commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/DeleteOrphanFilesSparkAction.java:
```diff
@@ -335,7 +344,39 @@ private Dataset<String> listedFileDS() {
+      if (!pathFilter.accept(currentPath)) {
+        isHiddenPath = true;
+        break;
+      }
```
nit: newline

liziyan-lzy replied:
Fixed. Thanks!

pvary commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/DeleteOrphanFilesSparkAction.java:
```diff
@@ -335,7 +344,39 @@ private Dataset<String> listedFileDS() {
+    if (!dir.endsWith("/")) {
+      listPath = dir + "/";
+    }
+    Iterable<org.apache.iceberg.io.FileInfo> files = io.listPrefix(listPath);
```
Note: if the filter removes full directories. The performance of the new listing could degrade

liziyan-lzy replied:
Thanks for highlighting this concern. I share the same worry. Since SupportsPrefixOperations currently only supports full recursive listing, full directory-level filtering could indeed cause performance issues. Perhaps we should consider adding depth support to SupportsPrefixOperations later？

pvary replied:
Maybe only use the listing when the filter is empty?

liziyan-lzy replied:
Just to confirm the behavior:
New listing for empty filters
Original listing when filters exist
Does this align with what you're expecting? I agree with this approach - it should address this performance regression issue. Thanks.

pvary replied:
We still will have performance issues if the `FileIO` implements `SupportsPrefixOperations` in a non-performant way. In this case instead of distributing the listing (as the original code does) we do it on a single node. (Note that the original code lists files/directories which are below the `MAX_DRIVER_LISTING_DEPTH`, then lists directories between `MAX_DRIVER_LISTING_DEPTH` and `MAX_DRIVER_LISTING_DIRECT_SUB_DIRS` and lists them using a distributed way.)

I think the original authors of the code done this to ensure timely listing for huge tables. So I think simply changing the behavior would cause issues in their use-cases. We can do performance tests to show that the listings for the current `FileIO` implementations are performant enough to get rid of this optimization. Whatever we find, we need to discuss this on the dev list before moving forward.

Once upon a time, long time ago, there were some discussions to make the file listing pluggable: https://github.com/apache/iceberg/issues/4346#issuecomment-1083412195

I'm not sure what the guys mentioned that think about it now. This might be another option to discuss on the dev list if you see compelling reasons to accept the complexity it brings.

liziyan-lzy replied:
I agree with you, this is very important. The current new listing method lost the capability for distributed listing, which can lead to performance regression when dealing with large numbers of files. I've been considering this issue as well.

Perhaps this implementation should not be the default method, or we could introduce a strategy approach, making the new listing method optional. This would also provide a solution for the issue mentioned in https://github.com/apache/iceberg/issues/11541. What do you think?

pvary replied:
They are discussing the exact same issues I highlighted above. Dan said that he has some ideas and plan to comment on the PR, but I can't find it.

I think we should ask on the dev list, but I expect that the prefix based listing is only viable for relatively small tables. When the number of files starts to scale up, the listing will become problematic.

liziyan-lzy replied:
Yes. Agree with you. I will bring this to the dev list to gather more insights. Thanks.

pvary commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/DeleteOrphanFilesSparkAction.java:
```diff
@@ -372,7 +413,7 @@ private static void listDirRecursively(
 
       for (String subDir : subDirs) {
-        listDirRecursively(
+        listDirRecursivelyWithHadoop(
```
Why is this `WithHadoop`?

liziyan-lzy replied:
Because this method is based on Hadoop FileSystem to list file status. Do you have any suggestions for method naming?

pvary replied:
Why did you decide to use the Hadoop listing here, and when do you start using the new listing?

liziyan-lzy replied:
Execution will enter this branch when FileIO does not implement SupportsPrefixOperations, so I maintain the existing behavior. Once FileIO supports SupportsPrefixOperations, the new listing logic will take effect. Do you have any suggestions on this? Thanks.

liziyan-lzy commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/DeleteOrphanFilesSparkAction.java:
```diff
@@ -335,7 +344,39 @@ private Dataset<String> listedFileDS() {
+        break;
+      }
+      currentPath = currentPath.getParent();
+    }
```
Fixed. Thanks!

fuzing commented:
@RussellSpitzer - We've applied this PR and performed some cursory testing with a minio S3 compatible store.

We scattered a number of random files inside and outside the table's sub-folders (including both the metadata and data folders/sub-folders) and it appears to be working correctly.

One anomaly (to be reconfirmed during tomorrow's testing run) is that we didn't see the expected output (i.e. the orphan_file_output array was empty - not helpful during our dry_run).

We used the following call:
```
    CALL system.remove_orphan_files(
    table => 'a.b.c',
    older_than => TIMESTAMP '2025-04-22T00:00:00.000Z',
    dry_run => true);
```

With dry_run => false for the removal.   As expected, extraneous files within s3://bucket/a/b were removed.  Those outside this path were left untouched.

Tomorrow we plan to confirm the existence/non-existence of the output files, along with injecting some orphaned files that pre-date/post-date the older_than timestamp to make sure that this is behaving as expected.

fuzing commented:
@RussellSpitzer - we ran further tests today and are pleased to report that for S3 this PR appears to work correctly with 
the following caveat:
- The files deleted (or proposed deletes in dry_run mode) don't seem to be populated in the procedure response payload.

Our tests involved injecting random files within and outside the Iceberg table S3 root prefix.  We injected files with various timestamps and noted that these faux-orphaned files under the root table prefix older than our "older_than" setting were removed correctly, while those younger were left intact.  No files outside our root table prefix were removed.

We're pretty sure there's a bug in file enumeration for the purposes of reporting (i.e. dry_run and actual removal reporting), in that they are not returned as part of the response payload through (in our case) the ODBC driver (whereas other query results, such as SQL SELECT results are).

If the above is deemed to be a bug, then this should be fixed, at which point this P/R should be ready to merge.  We're happy to rerun tests if needed.

Note that we did not test the fallback path (we aren't configured for such a setup unfortunately - we're Iceberg => S3 only).

liziyan-lzy commented:
> @RussellSpitzer - we ran further tests today and are pleased to report that for S3 this PR appears to work correctly with the following caveat:
> 
> * The files deleted (or proposed deletes in dry_run mode) don't seem to be populated in the procedure response payload.
> 
> Our tests involved injecting random files within and outside the Iceberg table S3 root prefix. We injected files with various timestamps and noted that these faux-orphaned files under the root table prefix older than our "older_than" setting were removed correctly, while those younger were left intact. No files outside our root table prefix were removed.
> 
> We're pretty sure there's a bug in file enumeration for the purposes of reporting (i.e. dry_run and actual removal reporting), in that they are not returned as part of the response payload through (in our case) the ODBC driver (whereas other query results, such as SQL SELECT results are).
> 
> If the above is deemed to be a bug, then this should be fixed, at which point this P/R should be ready to merge. We're happy to rerun tests if needed.
> 
> Note that we did not test the fallback path (we aren't configured for such a setup unfortunately - we're Iceberg => S3 only).

Thanks a lot for running those tests and spotting the issue. I'll dive into the file enumeration problem and get it resolved. Really appreciate your help with this!

AndreaPiccione commented:
It would be great to have this out soon. We are currently backfilling our datalake and are accumulating a large amount of data/metadata files. We have always performed vacuuming via Athena, but with the increase amount of data we are now hitting the 30-minute timeout.
Switching to doing rewriting, snapshot expiration and orphan removal via Spark will make our lives easier.

RussellSpitzer commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/DeleteOrphanFilesSparkAction.java:
```diff
@@ -302,21 +303,29 @@ private Dataset<FileURI> actualFileIdentDS() {
   private Dataset<String> listedFileDS() {
     List<String> subDirs = Lists.newArrayList();
     List<String> matchingFiles = Lists.newArrayList();
+    boolean isSupportPrefixOperations = table.io() instanceof SupportsPrefixOperations;
```
should this be "ioSupportsPrefixOperations"?

liziyan-lzy replied:
Yes. Thanks for pointing out.

pvary commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/DeleteOrphanFilesSparkAction.java:
```diff
@@ -206,6 +208,11 @@ public DeleteOrphanFilesSparkAction compareToFileList(Dataset<Row> files) {
     return this;
   }
 
+  public DeleteOrphanFilesSparkAction usePrefixList(boolean newUsePrefixList) {
```
nit: maybe `usePrefixListing`, or add javadoc. I think it is not trivial to understand what this config does

pvary commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/DeleteOrphanFilesSparkAction.java:
```diff
@@ -303,39 +310,88 @@ private Dataset<String> listedFileDS() {
+    }
+  }
+
+  private static boolean isHiddenPath(String baseDir, Path path, PathFilter pathFilter) {
```
Do we have CRC files, or something which might not be part of the table metadata, but needed by the FS?

liziyan-lzy replied:
Thanks for pointing out. These files typically appear as hidden and are ignored in both the old and new list file processes. However, if there are any specific concerns or edge cases related to handling these hidden files, please let me know. Thanks.

pvary replied:
Oh.. I missed that the filter already contains the `PartitionAwareHiddenPathFilter`. Sorry

pvary commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRemoveOrphanFilesAction.java:
```diff
@@ -107,9 +107,16 @@ public abstract class TestRemoveOrphanFilesAction extends TestBase {
+  @Parameters(name = "formatVersion = {0}, usePrefixList = {1}")
   protected static List<Object> parameters() {
-    return Arrays.asList(2, 3);
+    return Arrays.asList(
+        new Object[] {2, true},
+        new Object[] {3, true},
+        new Object[] {2, false},
+        new Object[] {3, false});
```
Do we need a full set of tests? These tests are costly, and running them might not be needed

liziyan-lzy replied:
Thank you for highlighting this. I have skipped necessary duplicate test runs covering both prefix listing and Hadoop file listing branch.

pvary replied:
Why did we keep the tests on both V3, and V2? Do we expect any differences?

liziyan-lzy replied:
Thank you for identifying this. There are no differences between V3 and V2 regarding this matter, so running the prefix listing tests exclusively on V2 should be sufficient. Thanks.

pvary commented on docs/docs/spark-procedures.md:
```diff
@@ -306,17 +306,18 @@ Used to remove files which are not referenced in any metadata files of an Iceber
+| `equal_schemes`         |    | map<string, string> | Mapping of file system schemes to be considered equal. Key is a comma-separated list of schemes and value is a scheme (defaults to `map('s3a,s3n','s3')`).                                 |
+| `equal_authorities`     |    | map<string, string> | Mapping of file system authorities to be considered equal. Key is a comma-separated list of authorities and value is an authority.                                                         |
+| `prefix_mismatch_mode`  |    | string | Action behavior when location prefixes (schemes/authorities) mismatch: <ul><li>ERROR - throw an exception. (default) </li><li>IGNORE - no action.</li><li>DELETE - delete files.</li></ul> |  
+| `prefix_list`           |    | boolean   | When true, use prefix-based file listing via the `SupportsPrefixOperations` interface. The Table FileIO implementation must support `SupportsPrefixOperations` when this flag is enabled (defaults to false) |
```
nit: maybe rename this to `prefix_listing` too?

liziyan-lzy replied:
Renamed to `prefix_listing`. Thanks!

pvary commented on docs/docs/spark-procedures.md:
```diff
@@ -306,17 +306,18 @@ Used to remove files which are not referenced in any metadata files of an Iceber
-| `equal_authorities` |    | map<string, string> | Mapping of file system authorities to be considered equal. Key is a comma-separated list of authorities and value is an authority. |
-| `prefix_mismatch_mode` |    | string | Action behavior when location prefixes (schemes/authorities) mismatch: <ul><li>ERROR - throw an exception. (default) </li><li>IGNORE - no action.</li><li>DELETE - delete files.</li></ul> |  
+| Argument Name           | Required? | Type | Description                                                                                                                                                                                |
+|-------------------------|-----------|------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `table`                 | ✔️  | string | Name of the table to clean                                                                                                                                                                 |
+| `older_than`            | ️   | timestamp | Remove orphan files created before this timestamp (Defaults to 3 days ago)                                                                                                                 |
+| `location`              |    | string    | Directory to look for files in (defaults to the table's location)                                                                                                                          |
+| `dry_run`               |    | boolean   | When true, don't actually remove files (defaults to false)                                                                                                                                 |
+| `max_concurrent_deletes` |    | int       | Size of the thread pool used for delete file actions (by default, no thread pool is used)                                                                                                  |
+| `file_list_view`        |    | string | Dataset to look for files in (skipping the directory listing)                                                                                                                              |
+| `equal_schemes`         |    | map<string, string> | Mapping of file system schemes to be considered equal. Key is a comma-separated list of schemes and value is a scheme (defaults to `map('s3a,s3n','s3')`).                                 |
+| `equal_authorities`     |    | map<string, string> | Mapping of file system authorities to be considered equal. Key is a comma-separated list of authorities and value is an authority.                                                         |
+| `prefix_mismatch_mode`  |    | string | Action behavior when location prefixes (schemes/authorities) mismatch: <ul><li>ERROR - throw an exception. (default) </li><li>IGNORE - no action.</li><li>DELETE - delete files.</li></ul> |  
```
Is this just a reformat? We usually don't reformat the table to avoid confusion

liziyan-lzy replied:
Got it! Reverted the reformatting to avoid confusion. Thanks for pointing this out.

pvary commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRemoveOrphanFilesAction.java:
```diff
@@ -384,7 +412,11 @@ public void testMetadataFolderIsIntact() {
-        actions.deleteOrphanFiles(table).olderThan(System.currentTimeMillis()).execute();
+        actions
+            .deleteOrphanFiles(table)
+            .usePrefixListing(usePrefixListing)
```
Do we have a test which checks the default value and bahaviour?

liziyan-lzy replied:
There are some tests do not need to be executed on both the prefix listing and Hadoop based listing branches. These tests will utilized default setting value and check default behavior.

pvary replied:
We usually prefer to have explicit test cases.
This kind of testing easily end up accidentally removed.

What do you think about adding a single tests with `testDefaultToHadoopListing`?

liziyan-lzy replied:
I agree with your suggestion. I have added a test `testDefaultToHadoopListing` to ensure default behavior. Thank you for your input!

pvary commented:
@RussellSpitzer: Any comments on the PR? If you are ok with this, I can merge

RussellSpitzer commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/DeleteOrphanFilesSparkAction.java:
```diff
@@ -303,39 +310,89 @@ private Dataset<String> listedFileDS() {
+    if (usePrefixListing) {
+      Preconditions.checkArgument(
+          table.io() instanceof SupportsPrefixOperations,
+          "Table file io should prefix operations when using prefix list.");
```
```suggestion
          "Cannot use prefix listing with FileIO {} which does not support prefix operations.", table.io());
```

Please double check ^ but we need to at least fix the grammar in the original message

RussellSpitzer commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/DeleteOrphanFilesSparkAction.java:
```diff
@@ -303,39 +310,89 @@ private Dataset<String> listedFileDS() {
+
+    Iterable<org.apache.iceberg.io.FileInfo> files = io.listPrefix(listPath);
+    for (org.apache.iceberg.io.FileInfo file : files) {
+      Path path = new Path(file.location());
```
A note to watch this space later since we are implicitly resolving paths here as Hadoop paths which may not be the right thing to do in S3 etc ... This is the same behavior though that the action would have had before so we can maybe do a followup on this if it's an issue.

RussellSpitzer commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRemoveOrphanFilesAction.java:
```diff
@@ -107,9 +115,12 @@ public abstract class TestRemoveOrphanFilesAction extends TestBase {
+  @Parameters(name = "formatVersion = {0}, usePrefixListing = {1}")
   protected static List<Object> parameters() {
-    return Arrays.asList(2, 3);
+    return Arrays.asList(new Object[] {2, true}, new Object[] {2, false}, new Object[] {3, false});
```
Why is format version 3 only tested with false? Also since we are changing this can you change this enumeration to use TestHelpers AllVersions

pvary replied:
I have explicitly asked to decrease the number of tests, as these tests are costly, and the listing charge has nothing to do with the format version.

There could be arguments for any version of testing matrix, but if we spend too much time on it, then we should just enable the whole matrix and wait a bit more for the CI to finish  😉

RussellSpitzer replied:
I mostly just don't want to have another explicit "version number" I'm fine if we even only test on the default version number really.

liziyan-lzy replied:
Thank you all for the discussion. I've updated the implementation to include all versions from TestHelpers.ALL_VERSIONS. The current test matrix maintains coverage for both prefix listing enabled and disabled cases, while optimizing execution by skipping redundant test cases from previous commits.

RussellSpitzer commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRemoveOrphanFilesAction.java:
```diff
@@ -344,6 +374,8 @@ public void testWapFilesAreKept() {
     // wap write
     df.select("c1", "c2", "c3").write().format("iceberg").mode("append").save(tableLocation);
 
+    spark.conf().unset(SparkSQLProperties.WAP_ID);
```
Why is this required?

RussellSpitzer replied:
We should be using withSQLConf then for this test then like

```java
    withSQLConf(
        ImmutableMap.of(confName, "2m"),
        () -> {
          SparkConfParser parser = new SparkConfParser(spark, table, ImmutableMap.of());
          Duration duration = parser.durationConf().sessionConf(confName).parseOptional();
          assertThat(duration).hasMinutes(2);
        });
 ```
 
 or a cloned Spark Session

liziyan-lzy replied:
Fixed. Used withSQLConf as suggested. Thanks for your guidance!

RussellSpitzer commented:
The only real fix I need is the error message, I have some other suggestions there to look into but they are more nits

liziyan-lzy commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRemoveOrphanFilesAction.java:
```diff
@@ -1067,6 +1148,38 @@ public void testRemoveOrphanFileActionWithDeleteMode() {
+  @TestTemplate
+  public void testDefaultToHadoopListing() {
+    assumeThat(usePrefixListing)
+        .as("Should not test both prefix listing and Hadoop file listing (redundant)")
```
Fixed. Thanks for pointing this out.

liziyan-lzy commented:
Hi @pvary , I've noticed there are some conflicts in this PR. Would you recommend that I rebase onto the current main branch and resolve conflicts?
Also, I see Spark 4.0 is now the default version. Should we keep these changes focused on the current target version and create a follow-up PR for Spark 4.0 compatibility？
Thanks for your guidance!

pvary commented:
> Hi @pvary , I've noticed there are some conflicts in this PR. Would you recommend that I rebase onto the current main branch and resolve conflicts?

Definitely do it please. We can't merge with conflicts 😄 

> Also, I see Spark 4.0 is now the default version. Should we keep these changes focused on the current target version and create a follow-up PR for Spark 4.0 compatibility？ Thanks for your guidance!

Since rebase is needed anyway, you might be able to move the changes to Spark 4.0. If you find it difficult, we can merge this to Spark 3.5, and then port the changes to the other supported versions.

We generally try to work on the latest version, so the changes slowly make to all of the supported versions. Even if we forget to backport the changes. But this one is only the backup plan. The work is finished on a feature when it is ported to every supported version.

liziyan-lzy commented:
> > Hi @pvary , I've noticed there are some conflicts in this PR. Would you recommend that I rebase onto the current main branch and resolve conflicts?
> 
> Definitely do it please. We can't merge with conflicts 😄
> 
> > Also, I see Spark 4.0 is now the default version. Should we keep these changes focused on the current target version and create a follow-up PR for Spark 4.0 compatibility？ Thanks for your guidance!
> 
> Since rebase is needed anyway, you might be able to move the changes to Spark 4.0. If you find it difficult, we can merge this to Spark 3.5, and then port the changes to the other supported versions.
> 
> We generally try to work on the latest version, so the changes slowly make to all of the supported versions. Even if we forget to backport the changes. But this one is only the backup plan. The work is finished on a feature when it is ported to every supported version.

Hi, I have resolved the merge conflicts. Would it make sense to first land this in Spark 3.5, then create a subsequent PR to port these changes to our other supported version branches?

pvary commented:
Merged to main.
Thanks @liziyan-lzy for the PR, and @RussellSpitzer for the review!

pvary commented:
@liziyan-lzy: Please port the changes to Spark 4.0 and 3.4

liziyan-lzy commented:
Thanks @pvary and @RussellSpitzer for the review! I'll create a follow-up PR to port these changes to Spark 4.0 and 3.4.

astronautas commented:
@AndreaPiccione Does it work for "gs://" scheme as well? We use PySpark 4.0.1 + Icebreg 1.10.1, and it does not work (shows scheme unsupported :/).
